File: jar-shell/src/main/java/com/beancounter/shell/ingest/AbstractIngester.java
package com.beancounter.shell.ingest;

import com.beancounter.client.services.PortfolioServiceClient;
import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.input.ImportFormat;
import com.beancounter.common.input.TrustedTrnImportRequest;
import com.beancounter.common.model.CallerRef;
import com.beancounter.common.model.Portfolio;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.SneakyThrows;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public abstract class AbstractIngester implements Ingester {
  private final Map<String, TrnWriter> writers = new HashMap<>();
  private PortfolioServiceClient portfolioService;

  @Autowired
  public void setPortfolioService(PortfolioServiceClient portfolioService) {
    this.portfolioService = portfolioService;
  }

  @Autowired
  public void setTrnWriters(TrnWriter... trnWriter) {
    for (TrnWriter writer : trnWriter) {
      String key = "DEFAULT";
      if (writer.id() != null) {
        key = writer.id().toUpperCase();
      }
      writers.put(key.toUpperCase(), writer);
    }
  }

  private TrnWriter getWriter(String id) {
    return writers.get(id.toUpperCase());
  }

  /**
   * Default ingestion flow.
   *
   * @param ingestionRequest parameters to run the import.
   */
  @SneakyThrows
  public void ingest(IngestionRequest ingestionRequest) {
    Portfolio portfolio = portfolioService.getPortfolioByCode(ingestionRequest.getPortfolioCode());
    TrnWriter writer = getWriter(ingestionRequest.getWriter());
    if (writer == null) {
      throw new BusinessException(
          String.format("Unable to resolve the Writer %s", ingestionRequest.getWriter())
      );
    }

    prepare(ingestionRequest, writer);
    List<List<String>> rows = getValues();
    int i = 0;
    for (List<String> row : rows) {
      CallerRef callerRef = new CallerRef(
          ingestionRequest.getProvider() == null
              ? portfolio.getId() : ingestionRequest.getProvider(),
          String.valueOf(i), String.valueOf(i++));

      TrustedTrnImportRequest trnRequest = new TrustedTrnImportRequest(
          portfolio, ImportFormat.SHARESIGHT, callerRef, "", row);
      writer.write(trnRequest);
    }

    writer.flush();
  }

  public abstract void prepare(IngestionRequest ingestionRequest, TrnWriter trnWriter);

  public abstract List<List<String>> getValues();
}

File: jar-shell/src/main/java/com/beancounter/shell/ingest/AbstractIngester.kt
package com.beancounter.shell.ingest

import com.beancounter.client.services.PortfolioServiceClient
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.input.ImportFormat
import com.beancounter.common.input.TrustedTrnImportRequest
import com.beancounter.common.model.CallerRef
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
/**
 * Abstract ingestion capabilities.
 */
abstract class AbstractIngester : Ingester {
    private val writers: MutableMap<String, TrnWriter> = HashMap()
    private lateinit var portfolioService: PortfolioServiceClient

    @Autowired
    fun setPortfolioService(portfolioService: PortfolioServiceClient) {
        this.portfolioService = portfolioService
    }

    @Autowired
    fun setTrnWriters(vararg trnWriter: TrnWriter) {
        for (writer in trnWriter) {
            writers[writer.id().toUpperCase()] = writer
        }
    }

    private fun getWriter(id: String): TrnWriter? {
        return writers[id.toUpperCase()]
    }

    /**
     * Default ingestion flow.
     *
     * @param ingestionRequest parameters to run the import.
     */
    override fun ingest(ingestionRequest: IngestionRequest) {
        val portfolio = portfolioService.getPortfolioByCode(ingestionRequest.portfolioCode!!)
        val writer = getWriter(ingestionRequest.writer)
            ?: throw BusinessException(String.format("Unable to resolve the Writer %s", ingestionRequest.writer))
        prepare(ingestionRequest, writer)
        val rows = values
        for ((i, row) in rows.withIndex()) {
            val callerRef = CallerRef(
                ingestionRequest.provider ?: portfolio.id, i.toString(), i.toString()
            )
            val trnRequest = TrustedTrnImportRequest(
                portfolio, ImportFormat.SHARESIGHT, callerRef, "", row
            )
            writer.write(trnRequest)
        }
        writer.flush()
    }

    abstract fun prepare(ingestionRequest: IngestionRequest, trnWriter: TrnWriter)
    abstract val values: List<List<String>>
}
